README.rst
Aegea: Amazon Web Services Operator Interface
=============================================

*Aegea* is a command line interface (CLI) that provides a set of essential commands and terminal dashboards for
operators of Amazon Web Services (AWS) accounts. Aegea lets you build AMIs and Docker images using the
`cloud-init <http://cloudinit.readthedocs.io/>`_ config management package, manage config roles, launch and monitor instances
and services, and manage AWS resources including ELB, RDS, and AWS Batch. It is intended to be used in conjunction with the
existing functionality of the `AWS CLI <https://aws.amazon.com/cli/>`_ and `boto3 <https://boto3.readthedocs.io/>`_.

Installation
~~~~~~~~~~~~
::

   pip install aegea

Before you do this, you will also need to install some system library dependencies:

+--------------+---------+-------------------------------------------------------------------------------------------------------+
| OS           | Python  | Command                                                                                               |
+==============+=========+=======================================================================================================+
| OS X         |         | ``sudo easy_install-2.7 pip; sudo pip install --ignore-installed six python-dateutil;``               |
|              |         | ``sudo pip install aegea`` (if prompted to install XCode Command Line Tools, do it and try again)     |
+--------------+---------+-------------------------------------------------------------------------------------------------------+
| Ubuntu       | Python 2| sudo apt-get update;                                                                                  |
|              |         | sudo apt-get install build-essential python-pip python-dev python-cffi libffi-dev libssl-dev moreutils|
+--------------+---------+-------------------------------------------------------------------------------------------------------+
| Ubuntu       | Python 3| sudo apt-get update;                                                                                  |
|              |         | sudo apt-get install build-essential python3-{pip,dev,cffi} libffi-dev libssl-dev moreutils           |
+--------------+---------+-------------------------------------------------------------------------------------------------------+
| Red Hat      | Python 2| sudo yum install python-devel python-cffi openssl-devel moreutils                                     |
+--------------+---------+-------------------------------------------------------------------------------------------------------+
| Red Hat      | Python 3| sudo yum install python3-devel python3-cffi openssl-devel moreutils                                   |
+--------------+---------+-------------------------------------------------------------------------------------------------------+

Run ``aws configure`` to configure `IAM <https://aws.amazon.com/iam/>`_ access credentials that will be used by the
``aws`` and ``aegea`` commands. You can create a new IAM key at https://console.aws.amazon.com/iam/home#/users.

**Ubuntu 12.04**: Use ``pip install cffi`` instead of ``apt-get install python-cffi``. Update your Python packaging utilities:
``for p in six setuptools packaging pip setuptools; do pip install --upgrade $p; hash -r; done``.

**Local install**: run ``make install`` in this directory.

**No root access; user-local install**: Use ``make install_venv`` to install aegea in its own virtualenv. The last line of the 
output shows how to activate the virtualenv. The version of virtualenv packaged in Ubuntu 12.04 is too old; use
``pip install --upgrade --user virtualenv`` to upgrade it.

**OS X with Homebrew Python**: Run ``brew link --force openssl``.

Aegea Batch
~~~~~~~~~~~
The `AWS Batch <https://aws.amazon.com/batch>`_ API currently requires you to use the us-east-1 region. You can use
``aws configure`` to select this region.

`aegea/missions/docker-example/ <aegea/missions/docker-example/>`_ is a root directory of an **aegea mission** -
a configuration management role. It has a rootfs.skel and a config.yml, which has directives to install packages,
etc. The example just installs the bwa APT package.

Run ``aegea-build-image-for-mission docker-example dex`` to build an ECR image called dex from the "docker-example"
mission. You can list ECR images with ``aegea ecr ls``, and delete them with e.g. ``aws ecr delete-repository dex``.

Run ``aegea batch submit --ecs-image dex --command "bwa aln || true" "bwa mem || true" --memory 2048 --vcpus 4 --watch``
to run a Batch job that requires 2 GB RAM and 4 cores to be allocated to the Docker container, using the "dex" image,
and executes two commands as listed after --command, using "bash -euo pipefail -c".

You can also use ``aegea batch submit --execute FILE``. This will slurp up FILE (any type of shell script or ELF
executable) and execute it in the job's Docker container.

The concurrency and cost of your Batch jobs is governed by the "Max vCPUs" setting in your compute environment.
To change the capacity or other settings of your compute environment, go to
https://console.aws.amazon.com/batch/home?region=us-east-1#/compute-environments, select "aegea_batch", and click "Edit".

AWS Batch launches and manages `ECS <https://aws.amazon.com/ecs/>`_ host instances to execute your jobs. You can see the
host instances by running ``aegea ls``.

Configuration management
~~~~~~~~~~~~~~~~~~~~~~~~
Aegea supports ingesting configuration from a configurable array of sources. Each source is a JSON or YAML file.
Configuration sources that follow the first source update the configuration using recursive dictionary merging. Sources are
enumerated in the following order (i.e., in order of increasing priority):

- Site-wide configuration source, ``/etc/aegea/config.yml``
- User configuration source, ``~/.config/aegea/config.yml``
- Any sources listed in the colon-delimited variable ``AEGEA_CONFIG_FILE``
- Command line options

**Array merge operators**: When loading a chain of configuration sources, Aegea uses recursive dictionary merging to
combine the sources. Additionally, when the original config value is a list, Aegea supports array manipulation
operators, which let you extend and modify arrays defined in underlying configurations. See
https://github.com/kislyuk/tweak#array-merge-operators for a list of these operators.

Building AMIs and Docker images
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Aegea includes a lightweight configuration management system for building machine images based on
`cloud-init <http://cloudinit.readthedocs.io/>`_ (both Docker images and AMIs are supported).

TODO: ``build_image build_ami build_docker_image rootfs.skel``

.. image:: https://circleci.com/gh/kislyuk/aegea.svg?style=svg&circle-token=70d22b84025fad5d484ac5f3df1fc0a183c0f516
   :target: https://circleci.com/gh/kislyuk/aegea
.. image:: https://codecov.io/gh/kislyuk/aegea/branch/master/graph/badge.svg?token=a9suCNpECz
   :target: https://codecov.io/gh/kislyuk/aegea


Changes.rst
Changes for v1.9.18 (2017-08-16)
================================

-  aegea batch watch: Do not crash if log stream does not exist yet

Changes for v1.9.17 (2017-06-15)
================================

Merge pull request #22 from wholebiome/build-timeout Extend timeout for
AMI builds Added timeout to loop, default much longer Fix tests

Changes for v1.9.16 (2017-06-01)
================================

-  Add file missed in 0c99863

Changes for v1.9.15 (2017-06-01)
================================

-  Fix logic error in parameter naming

Changes for v1.9.14 (2017-05-29)
================================

-  Temporarily disable batch custom AMI

Changes for v1.9.13 (2017-05-29)
================================

-  Minor refactor in batch

-  Ensure default selection of batch instances has instance storage

-  Begin aegea lambda ls, aegea rm --lambda

-  Tab complete log levels

-  Avoid using pkgutil for introspection

Changes for v1.9.12 (2017-05-14)
================================

-  Batch bug fixes and begin support for custom Batch ECI AMIs

Changes for v1.8.4 (2017-02-02)
===============================

-  Install process robustness improvements

-  Install documentation improvements

Changes for v1.8.3 (2017-02-01)
===============================

-  Don't symlink aegea in bin to avoid pip uninstall bugs

Changes for v1.8.2 (2017-02-01)
===============================

-  Resume interrupted release

Changes for v1.8.1 (2017-02-01)
===============================

-  Resume interrupted release

Changes for v1.8.0 (2017-02-01)
===============================

-  Installation documentation and robustness improvements

-  Batch API and mission-specific improvements

Changes for v1.7.4 (2017-01-26)
===============================

-  aegea batch: automatic setup of builder IAM policies

-  aegea batch submit --job-role: automatic setup of job IAM roles

-  aegea batch submit --storage: EBS volume manager

-  Autocomplete column titles in listing subcommands where a resource is
   available

-  Autoconfigure a VPC if all VPCs including the default VPC were
   deleted

-  Asset loader: offload rootfs.skel to S3 when user-data exceeds 16K
   limit

-  Arvados updates

-  Make missions dir doc link relative (#9)

-  Display statusReason in aegea batch ls and aegea batch watch

Changes for v1.7.3 (2017-01-18)
===============================

-  Add automatic configuration for route53 private DNS

-  Various improvements to aegea batch

-  Work around autoloader import issue seen on some Python 2.7 versions

-  aegea build\_ami: improve progress and error messages

Changes for v1.7.2 (2017-01-13)
===============================

-  Fix makefile shell assumption

-  Batch WIP

Changes for v1.7.1 (2017-01-13)
===============================

-  Test and release infra improvements

-  Batch docs

Changes for v1.7.0 (2017-01-10)
===============================

-  aegea-build-image-for-mission now builds ECR images by default

-  Integration work for Batch

Changes for v1.6.3 (2017-01-08)
===============================

-  Add ELB SG configurator, aegea-rebuild-public-elb-sg

-  Add awscli to deps

Changes for v1.6.2 (2017-01-06)
===============================

-  ELB deploy: set default target group name properly

-  Make sure wheel is installed before attempting setup

-  Aegea batch submit: Begin CWL support

-  Aegea batch watch: amend log line dup fix

Changes for v1.6.1 (2017-01-03)
===============================

-  Improvements to aegea batch

Changes for v1.6.0 (2016-12-30)
===============================

-  Aegea EFS refactor

-  Aegea batch

-  Add IP Ranges API

-  Add aegea buckets cors placeholder

-  Aegea bucket lifecycle

-  Test and release infrastructure improvements

Changes for v1.5.1 (2016-11-14)
===============================

-  Fogdog mission: add environment placeholder

-  Begin timestamp backport

-  Propagate base AMI metadata in build\_image

Changes for v1.5.0 (2016-11-10)
===============================

-  Implement aegea rds snapshot

-  Only use pager with pretty-printed tables

-  Add Amazon Linux AMI locator

-  Use -w0 for auto col width table formatter

-  aegea zones update: support multiple updates

-  Cosmetic and documentation fixes

Changes for v1.4.0 (2016-11-02)
===============================

-  aegea-build-ami-for-mission: skip make if no Makefile
-  Begin FogDog mission
-  Arvados config support; improve config file handling
-  Don't fail cloud-init on account of expected ssh failure
-  Run ssh-add from aegea launch
-  aegea elb create bugfix
-  Fix ELB behavior when TG is present
-  Simplify arg forwarding in build\_ami

Changes for v1.3.0 (2016-10-20)
===============================

-  Support running core aegea on Ubuntu 14.04 vendored Python

-  Improve freeform cloud-config-data passing

-  Fix pager; introduce --auto-col-width table formatter

-  List security groups in elb listing

-  Break out and begin buildout of aegea ebs subcommand

-  Begin improving rds listings

-  Improve DNS zone repr

-  New protocol to check out local tracking branch in aegea deploy

-  aegea elb create: configurable health check path

-  Key cloud-init files manifest by file path to avoid duplicates

Changes for v1.2.2 (2016-10-08)
===============================

-  ELB provisioning and listing improvements

Changes for v1.2.1 (2016-10-07)
===============================

-  Aegea deploy fixups

Changes for v1.2.0 (2016-10-05)
===============================

-  Online documentation improvements


aegea/audit.py
from __future__ import absolute_import, division, print_function, unicode_literals

import os, sys, time, csv, json, unittest
from botocore.exceptions import ClientError
from datetime import datetime, timedelta
import dateutil.parser
from dateutil.tz import tzutc

from . import register_parser, logger, config
from .util import natural_sort
from .util.aws import expect_error_codes, ARN, clients, resources
from .util.printing import RED, GREEN, WHITE, page_output, format_table

class Auditor(unittest.TestCase):
    cache = {}
    def runTest(self):
        pass

    @property
    def credential_report(self):
        if "credential_report" not in self.cache:
            iam = clients.iam
            iam.generate_credential_report()
            while True:
                try:
                    self.cache["credential_report"] = iam.get_credential_report()
                    break
                except ClientError as e:
                    expect_error_codes(e, "ReportInProgress")
        return csv.DictReader(self.cache["credential_report"]["Content"].decode("utf-8").splitlines())

    @property
    def account_password_policy(self):
        if "account_password_policy" not in self.cache:
            self.cache["account_password_policy"] = resources.iam.AccountPasswordPolicy()
        return self.cache["account_password_policy"]

    @property
    def trails(self):
        if "trails" not in self.cache:
            self.cache["trails"] = clients.cloudtrail.describe_trails()["trailList"]
        return self.cache["trails"]

    @property
    def alarms(self):
        if "alarms" not in self.cache:
            self.cache["alarms"] = list(resources.cloudwatch.alarms.all())
        return self.cache["alarms"]

    def parse_date(self, d):
        return dateutil.parser.parse(d)

    def audit_1_1(self):
        """1.1 Avoid the use of the "root" account (Scored)"""
        for row in self.credential_report:
            if row["user"] == "<root_account>":
                for field in "password_last_used", "access_key_1_last_used_date", "access_key_2_last_used_date":
                    if row[field] != "N/A" and self.parse_date(row[field]) > datetime.now(tzutc()) - timedelta(days=1):
                        raise Exception("Root account last used less than a day ago ({})".format(field))

    def audit_1_2(self):
        """1.2 Ensure multi-factor authentication (MFA) is enabled for all IAM users that have a console password (Scored)"""  # noqa
        for row in self.credential_report:
            if row["user"] == "<root_account>" or json.loads(row["password_enabled"]):
                if not json.loads(row["mfa_active"]):
                    raise Exception("Account {} has a console password but no MFA".format(row["user"]))

    def audit_1_3(self):
        """1.3 Ensure credentials unused for 90 days or greater are disabled (Scored)"""
        for row in self.credential_report:
            for access_key in "1", "2":
                if json.loads(row["access_key_{}_active".format(access_key)]):
                    last_used = row["access_key_{}_last_used_date".format(access_key)]
                    if last_used != "N/A" and self.parse_date(last_used) < datetime.now(tzutc()) - timedelta(days=90):
                        msg = "Active access key {} in account {} last used over 90 days ago"
                        raise Exception(msg.format(access_key, row["user"]))

    def audit_1_4(self):
        """1.4 Ensure access keys are rotated every 90 days or less (Scored)"""
        for row in self.credential_report:
            for access_key in "1", "2":
                if json.loads(row["access_key_{}_active".format(access_key)]):
                    last_rotated = row["access_key_{}_last_rotated".format(access_key)]
                    if self.parse_date(last_rotated) < datetime.now(tzutc()) - timedelta(days=90):
                        msg = "Active access key {} in account {} last rotated over 90 days ago"
                        raise Exception(msg.format(access_key, row["user"]))

    def audit_1_5(self):
        """1.5 Ensure IAM password policy requires at least one uppercase letter (Scored)"""
        self.assertTrue(self.account_password_policy.require_uppercase_characters)

    def audit_1_6(self):
        """1.6 Ensure IAM password policy require at least one lowercase letter (Scored)"""
        self.assertTrue(self.account_password_policy.require_lowercase_characters)

    def audit_1_7(self):
        """1.7 Ensure IAM password policy require at least one symbol (Scored)"""
        self.assertTrue(self.account_password_policy.require_symbols)

    def audit_1_8(self):
        """1.8 Ensure IAM password policy require at least one number (Scored)"""
        self.assertTrue(self.account_password_policy.require_numbers)

    def audit_1_9(self):
        """1.9 Ensure IAM password policy requires minimum length of 14 or greater (Scored)"""
        self.assertGreaterEqual(self.account_password_policy.minimum_password_length, 14)

    def audit_1_10(self):
        """1.10 Ensure IAM password policy prevents password reuse (Scored)"""
        self.assertTrue(self.account_password_policy.password_reuse_prevention)

    def audit_1_11(self):
        """1.11 Ensure IAM password policy expires passwords within 90 days or less (Scored)"""
        self.assertLessEqual(self.account_password_policy.max_password_age, 90)

    def audit_1_12(self):
        """1.12 Ensure no root account access key exists (Scored)"""
        for row in self.credential_report:
            if row["user"] == "<root_account>":
                self.assertFalse(json.loads(row["access_key_1_active"]))
                self.assertFalse(json.loads(row["access_key_2_active"]))

    def audit_1_13(self):
        """1.13 Ensure hardware MFA is enabled for the "root" account (Scored)"""
        for row in self.credential_report:
            if row["user"] == "<root_account>":
                self.assertTrue(json.loads(row["mfa_active"]))

    def audit_1_14(self):
        """1.14 Ensure security questions are registered in the AWS account (Not Scored)"""

    def audit_1_15(self):
        """1.15 Ensure IAM policies are attached only to groups or roles (Scored)"""
        for policy in resources.iam.policies.all():
            self.assertEqual(len(list(policy.attached_users.all())), 0, "{} has users attached to it".format(policy))

    def audit_2_1(self):
        """2.1 Ensure CloudTrail is enabled in all regions (Scored)"""
        self.assertTrue(any(trail["IsMultiRegionTrail"] for trail in self.trails), "No multi-region CloudTrail trails")

    def audit_2_2(self):
        """2.2 Ensure CloudTrail log file validation is enabled (Scored)"""
        self.assertGreater(len(self.trails), 0, "No CloudTrail trails configured")
        self.assertTrue(all(trail["LogFileValidationEnabled"] for trail in self.trails),
                        "Some CloudTrail trails don't have log file validation enabled")

    def audit_2_3(self):
        """2.3 Ensure the S3 bucket CloudTrail logs to is not publicly accessible (Scored)"""
        raise NotImplementedError()
        import boto3
        s3 = boto3.session.Session(region_name="us-east-1").resource("s3")
        # s3 = boto3.resource("s3")
        # for trail in self.trails:
        #    for grant in s3.Bucket(trail["S3BucketName"]).Acl().grants:
        #    print(s3.Bucket(trail["S3BucketName"]).Policy().policy)
        for bucket in s3.buckets.all():
            print(bucket)
            try:
                print("    Policy:", bucket.Policy().policy)
            except:
                pass
            for grant in bucket.Acl().grants:
                try:
                    print("    Grant:", grant)
                except:
                    pass

    def audit_2_4(self):
        """2.4 Ensure CloudTrail trails are integrated with CloudWatch Logs (Scored)"""
        for trail in self.trails:
            self.assertIn("CloudWatchLogsLogGroupArn", trail)
            trail_status = clients.cloudtrail.get_trail_status(Name=trail["TrailARN"])
            self.assertGreater(trail_status["LatestCloudWatchLogsDeliveryTime"],
                               datetime.now(tzutc()) - timedelta(days=1))

    def audit_2_5(self):
        """2.5 Ensure AWS Config is enabled in all regions (Scored)"""
        import boto3
        for region in boto3.Session().get_available_regions("config"):
            aws_config = boto3.session.Session(region_name=region).client("config")
            res = aws_config.describe_configuration_recorder_status()
            self.assertGreater(len(res["ConfigurationRecordersStatus"]), 0)

    def audit_2_6(self):
        """2.6 Ensure S3 bucket access logging is enabled on the CloudTrail S3 bucket (Scored)"""
        raise NotImplementedError()

    def audit_2_7(self):
        """2.7 Ensure CloudTrail logs are encrypted at rest using KMS CMKs (Scored)"""
        raise NotImplementedError()

    def audit_2_8(self):
        """2.8 Ensure rotation for customer created CMKs is enabled (Scored)"""
        raise NotImplementedError()

    def ensure_alarm(self, name, pattern, log_group_name):
        # See http://docs.aws.amazon.com/AmazonCloudWatch/latest/DeveloperGuide/CW_Support_For_AWS.html
        sns = resources.sns
        logs = clients.logs
        cloudwatch = clients.cloudwatch
        topic = sns.create_topic(Name=name)
        topic.subscribe(Protocol='email', Endpoint=self.email)
        logs.put_metric_filter(logGroupName=log_group_name,
                               filterName=name,
                               filterPattern=pattern,
                               metricTransformations=[dict(metricName=name,
                                                           metricNamespace=__name__,
                                                           metricValue="1")])
        cloudwatch.put_metric_alarm(AlarmName=name,
                                    MetricName=name,
                                    Namespace=__name__,
                                    Statistic="Sum",
                                    Period=300,
                                    Threshold=1,
                                    ComparisonOperator="GreaterThanOrEqualToThreshold",
                                    EvaluationPeriods=1,
                                    AlarmActions=[topic.arn])

    def assert_alarm(self, name, pattern, remediate=False):
        logs = clients.logs
        sns = resources.sns
        alarm_ok = False
        for trail in self.trails:
            log_group_name = ARN(trail["CloudWatchLogsLogGroupArn"]).resource.split(":")[1]
            for metric_filter in logs.describe_metric_filters(logGroupName=log_group_name)["metricFilters"]:
                if metric_filter["filterPattern"] == pattern:
                    for alarm in self.alarms:
                        try:
                            self.assertEqual(alarm.metric_name, metric_filter["metricTransformations"][0]["metricName"])
                            self.assertGreater(len(list(sns.Topic(alarm.alarm_actions[0]).subscriptions.all())), 0)
                            alarm_ok = True
                        except Exception:
                            pass
        if remediate and not alarm_ok:
            self.ensure_alarm(name=name,
                              pattern=pattern,
                              log_group_name=log_group_name)
            alarm_ok = True
        self.assertTrue(alarm_ok)

    def audit_3_1(self):
        """3.1 Ensure a log metric filter and alarm exist for unauthorized API calls (Scored)"""
        self.assert_alarm("UnauthorizedAPICalls",
                          '{ ($.errorCode = "*UnauthorizedOperation") || ($.errorCode = "AccessDenied*") }')

    def audit_3_2(self):
        """3.2 Ensure a log metric filter and alarm exist for Management Console sign-in without MFA (Scored)"""
        self.assert_alarm("ConsoleUseWithoutMFA",
                          '{ $.userIdentity.sessionContext.attributes.mfaAuthenticated != "true" }')

    def audit_3_3(self):
        """3.3 Ensure a log metric filter and alarm exist for usage of "root" account (Scored)"""
        self.assert_alarm("RootAccountUsed",
                          '{ $.userIdentity.type = \"Root\" && $.userIdentity.invokedBy NOT EXISTS && $.eventType != \"AwsServiceEvent\" }')  # noqa

    def audit_3_4(self):
        """3.4 Ensure a log metric filter and alarm exist for IAM policy changes (Scored)"""
        self.assert_alarm("IAMPolicyChanged",
                          '{($.eventName=DeleteGroupPolicy)||($.eventName=DeleteRolePolicy)||($.eventName=DeleteUserPolicy)||($.eventName=PutGroupPolicy)||($.eventName=PutRolePolicy)||($.eventName=PutUserPolicy)||($.eventName=CreatePolicy)||($.eventName=DeletePolicy)||($.eventName=CreatePolicyVersion)||($.eventName=DeletePolicyVersion)||($.eventName=AttachRolePolicy)||($.eventName=DetachRolePolicy)||($.eventName=AttachUserPolicy)||($.eventName=DetachUserPolicy)||($.eventName=AttachGroupPolicy)||($.eventName=DetachGroupPolicy)}')  # noqa

    def audit_3_5(self):
        """3.5 Ensure a log metric filter and alarm exist for CloudTrail configuration changes (Scored)"""
        self.assert_alarm("CloudTrailConfigChanged",
                          '{ ($.eventName = CreateTrail) || ($.eventName = UpdateTrail) || ($.eventName = DeleteTrail) || ($.eventName = StartLogging) || ($.eventName = StopLogging) }')  # noqa

    def audit_3_6(self):
        """3.6 Ensure a log metric filter and alarm exist for AWS Management Console authentication failures (Scored)"""
        self.assert_alarm("ConsoleLoginFailed",
                          '{ ($.eventName = ConsoleLogin) && ($.errorMessage = \"Failed authentication\") }')

    def audit_3_7(self):
        """3.7 Ensure a log metric filter and alarm exist for disabling or scheduled deletion of customer created CMKs (Scored)"""  # noqa
        self.assert_alarm("KMSCMKDisabled",
                          '{($.eventSource = kms.amazonaws.com) && (($.eventName=DisableKey)||($.eventName=ScheduleKeyDeletion))}')  # noqa

    def audit_3_8(self):
        """3.8 Ensure a log metric filter and alarm exist for S3 bucket policy changes (Scored)"""
        self.assert_alarm("S3BucketPolicyChanged",
                          '{ ($.eventSource = s3.amazonaws.com) && (($.eventName = PutBucketAcl) || ($.eventName = PutBucketPolicy) || ($.eventName = PutBucketCors) || ($.eventName = PutBucketLifecycle) || ($.eventName = PutBucketReplication) || ($.eventName = DeleteBucketPolicy) || ($.eventName = DeleteBucketCors) || ($.eventName = DeleteBucketLifecycle) || ($.eventName = DeleteBucketReplication)) }')  # noqa

    def audit_3_9(self):
        """3.9 Ensure a log metric filter and alarm exist for AWS Config configuration changes (Scored)"""
        self.assert_alarm("AWSConfigServiceChanged",
                          '{($.eventSource = config.amazonaws.com) && (($.eventName=StopConfigurationRecorder)||($.eventName=DeleteDeliveryChannel)||($.eventName=PutDeliveryChannel)||($.eventName=PutConfigurationRecorder))}')  # noqa

    def audit_3_10(self):
        """3.10 Ensure a log metric filter and alarm exist for security group changes (Scored)"""
        self.assert_alarm("EC2SecurityGroupChanged",
                          '{ ($.eventName = AuthorizeSecurityGroupIngress) || ($.eventName = AuthorizeSecurityGroupEgress) || ($.eventName = RevokeSecurityGroupIngress) || ($.eventName = RevokeSecurityGroupEgress) || ($.eventName = CreateSecurityGroup) || ($.eventName = DeleteSecurityGroup)}')  # noqa

    def audit_3_11(self):
        """3.11 Ensure a log metric filter and alarm exist for changes to Network Access Control Lists (NACL) (Scored)"""  # noqa
        self.assert_alarm("EC2NACLChanged",
                          '{ ($.eventName = CreateNetworkAcl) || ($.eventName = CreateNetworkAclEntry) || ($.eventName = DeleteNetworkAcl) || ($.eventName = DeleteNetworkAclEntry) || ($.eventName = ReplaceNetworkAclEntry) || ($.eventName = ReplaceNetworkAclAssociation) }')  # noqa

    def audit_3_12(self):
        """3.12 Ensure a log metric filter and alarm exist for changes to network gateways (Scored)"""
        self.assert_alarm("EC2NetworkGatewayChanged",
                          '{ ($.eventName = CreateCustomerGateway) || ($.eventName = DeleteCustomerGateway) || ($.eventName = AttachInternetGateway) || ($.eventName = CreateInternetGateway) || ($.eventName = DeleteInternetGateway) || ($.eventName = DetachInternetGateway) }')  # noqa

    def audit_3_13(self):
        """3.13 Ensure a log metric filter and alarm exist for route table changes (Scored)"""
        self.assert_alarm("EC2RouteTableChanged",
                          '{ ($.eventName = CreateRoute) || ($.eventName = CreateRouteTable) || ($.eventName = ReplaceRoute) || ($.eventName = ReplaceRouteTableAssociation) || ($.eventName = DeleteRouteTable) || ($.eventName = DeleteRoute) || ($.eventName = DisassociateRouteTable) }')  # noqa

    def audit_3_14(self):
        """3.14 Ensure a log metric filter and alarm exist for VPC changes (Scored)"""
        self.assert_alarm("EC2VPCChanged",
                          '{ ($.eventName = CreateVpc) || ($.eventName = DeleteVpc) || ($.eventName = ModifyVpcAttribute) || ($.eventName = AcceptVpcPeeringConnection) || ($.eventName = CreateVpcPeeringConnection) || ($.eventName = DeleteVpcPeeringConnection) || ($.eventName = RejectVpcPeeringConnection) || ($.eventName = AttachClassicLinkVpc) || ($.eventName = DetachClassicLinkVpc) || ($.eventName = DisableVpcClassicLink) || ($.eventName = EnableVpcClassicLink) }')  # noqa

    def audit_3_15(self):
        """3.15 Ensure security contact information is registered (Scored)"""
        raise NotImplementedError()

    def audit_3_16(self):
        """3.16 Ensure appropriate subscribers to each SNS topic (Not Scored)"""
        raise NotImplementedError()

    def audit_4_1(self):
        """4.1 Ensure no security groups allow ingress from 0.0.0.0/0 to port 22 (Scored)"""
        raise NotImplementedError()

    def audit_4_2(self):
        """4.2 Ensure no security groups allow ingress from 0.0.0.0/0 to port 3389 (Scored)"""
        raise NotImplementedError()

    def audit_4_3(self):
        """4.3 Ensure VPC Flow Logging is Enabled in all Applicable Regions (Scored)"""
        raise NotImplementedError()

    def audit_4_4(self):
        """4.4 Ensure the default security group restricts all traffic (Scored)"""
        raise NotImplementedError()

def audit(args):
    auditor = Auditor()
    auditor.__dict__.update(vars(args))
    table = []
    for method_name in natural_sort(dir(auditor)):
        if method_name.startswith("audit"):
            method = getattr(auditor, method_name)
            try:
                method()
                table.append([GREEN("PASS"), method.__doc__])
            except Exception as e:
                logger.warn("%s: %s", method, e)
                table.append([RED("FAIL"), method.__doc__])
    # TODO: WHITE("NO TEST")
    page_output(format_table(table, column_names=["Result", "Test"], max_col_width=120))

parser = register_parser(audit, help='Generate a security report using the CIS AWS Foundations Benchmark')
parser.add_argument('--email', help="Administrative contact email")
